lib/ddtrace/metrics.rb

```diff
@@ -247,16 +256,18 @@ def gauge(stat, value, options = nil)
     INCOMPATIBLE_STATSD_ONLY_ONCE = Datadog::Utils::OnlyOnce.new
     private_constant :INCOMPATIBLE_STATSD_ONLY_ONCE
 
-    # IMPORTANT: Once you delete this, you probably want to update the `#ignored_statsd_warning` below where we
-    # recommend that customers add `gem 'dogstatsd-ruby', '~> 4.0'` to their Gemfile to perhaps state something
-    # different.
+    # Using dogstatsd-ruby >= 5.0 and < 5.2 can lead to unbounded memory usage in some scenario with process forks
+    # involved. Display a warning if the available version of the dogstatsd-ruby gem is impacted.
+    # Note that starting with 5.2.0, the `single_thread` mode of the dogstatsd-ruby gem is used in order to not have
+    # this issue.
     def incompatible_statsd_warning
-      return if Gem.loaded_specs['dogstatsd-ruby'].version < Gem::Version.new('5.0')
+      return if dogstatsd_gem_version < Gem::Version.new('5.0') || dogstatsd_gem_version >= Gem::Version.new('5.2')
 
       INCOMPATIBLE_STATSD_ONLY_ONCE.run do
         Datadog.logger.warn(
-          'This version of `ddtrace` is incompatible with `dogstastd-ruby` version >= 5.0 and can ' \
-          'cause unbounded memory usage. Please use `dogstastd-ruby` version < 5.0 instead.'
+          'This version of `ddtrace` is incompatible with `dogstastd-ruby` version >= 5.0 and < 5.2 and can '   \
+          'cause unbounded memory usage. Please use either `dogstastd-ruby` version < 5.0 or `dogstatsd-ruby` ' \
+          'version >= 5.2 instead.'
         )
```

Comment:
Is this "only warn but let users go ahead" behavior what we still want?

Previously, I could see how if users had previously migrated to directly use the 5.0 APIs forcing them to use a 4.x release would be annoying.

But now that there's a compatible 5.x release, should dd-trace-rb outright refuse to work with `dogstatsd-ruby` versions 5.0.0, 5.0.1, and 5.1.0? That way customers would never accidentally run into the issue anymore -- they either have a working dogstatsd version OR we disable the dogstatsd usage until they either upgrade to 5.2 or downgrade to 4.x.

Reply:
I would indeed prefer this behavior, I don't know if it is a common usage or if it is already something done in dd-trace-rb for another gem?

There is scenario where using 5.0, 5.0.1 or 5.1.0 won't lead to any issue, however, I don't see any reason to stick with these versions neither.

Reply:
I'd say so: on this very file, you'll notice that if you have a too old version of dogstatsd, we already just refuse to use it:

https://github.com/DataDog/dd-trace-rb/blob/461a627641755f4338e1e7dbb93d366d4534d968/lib/ddtrace/metrics.rb#L16-L23

Furthermore, we also [detect a few incompatible versions of gems and gracefully degrade the profiler](https://github.com/DataDog/dd-trace-rb/blob/master/lib/ddtrace/profiling/ext/cpu.rb#L43).

> There is scenario where using 5.0, 5.0.1 or 5.1.0 won't lead to any issue, however, I don't see any reason to stick with these versions neither.

Indeed! I don't think there's any good reason for customers to use any of those versions anymore, so I would suggest we go the safe way :)

Reply:
You're right.
Let me refactor this.

Reply:
Should be addressed

Reply:
👍 i think this approach makes sense as well